server/src/main/java/org/elasticsearch/action/support/RetryableAction.java

```diff
@@ -0,0 +1,136 @@
+
+    public RetryableAction(Logger logger, ThreadPool threadPool, TimeValue initialDelay, TimeValue timeoutValue,
+                           ActionListener<Response> listener) {
+        this(logger, threadPool, initialDelay, timeoutValue, listener, ThreadPool.Names.SAME);
+    }
+
+    public RetryableAction(Logger logger, ThreadPool threadPool, TimeValue initialDelay, TimeValue timeoutValue,
+                           ActionListener<Response> listener, String retryExecutor) {
+        this.logger = logger;
+        this.threadPool = threadPool;
+        this.initialDelayMillis = Math.max(initialDelay.getMillis(), 1);
+        this.timeoutMillis = Math.max(timeoutValue.getMillis(), 1);
+        this.startMillis = threadPool.relativeTimeInMillis();
+        this.finalListener = listener;
+        this.retryExecutor = retryExecutor;
+    }
+
+    public void run() {
+        runTryAction(new RetryingListener(initialDelayMillis, null));
+    }
+
+    public void cancel(Exception e) {
+        if (isDone.compareAndSet(false, true)) {
+            finalListener.onFailure(e);
+        }
+
+    }
+
+    private void runTryAction(ActionListener<Response> listener) {
+        try {
+            tryAction(listener);
+        } catch (Exception e) {
+            listener.onFailure(e);
+        }
+    }
+
+    public abstract void tryAction(ActionListener<Response> listener);
+
+    public abstract boolean shouldRetry(Exception e);
+
+    private class RetryingListener implements ActionListener<Response> {
+
+        private final long nextDelayMillis;
+        private final Exception existingException;
+
+        private RetryingListener(long nextDelayMillis, Exception existingException) {
+            this.nextDelayMillis = nextDelayMillis;
+            this.existingException = existingException;
+        }
+
+        @Override
+        public void onResponse(Response response) {
+            if (isDone.compareAndSet(false, true)) {
+                finalListener.onResponse(response);
+            }
+        }
+
+        @Override
+        public void onFailure(Exception e) {
+            if (shouldRetry(e)) {
+                final long elapsedMillis = threadPool.relativeTimeInMillis() - startMillis;
+                if (elapsedMillis > timeoutMillis) {
+                    logger.debug(() -> new ParameterizedMessage("retryable action timed out after {}",
+                        TimeValue.timeValueMillis(elapsedMillis)), e);
+                    addExisting(e);
+                    if (isDone.compareAndSet(false, true)) {
+                        finalListener.onFailure(e);
+                    }
+                } else {
+                    logger.debug(() -> new ParameterizedMessage("retrying action that failed in {}",
+                        TimeValue.timeValueMillis(nextDelayMillis)), e);
+                    addExisting(e);
+                    Runnable runnable = () -> runTryAction(new RetryingListener(nextDelayMillis * 2, e));
```

Comment:
let's use an ActionRunnable here that will also notify the listener if the action can't be enqueued on the threadpool of the retryExecutor